File: backend/internal/cli/format.go
package cli

import (
	"fmt"
	"strings"
	"time"
)

// formatBytes renders a byte count the way a download client should: three
// significant figures at most, so the number stops jittering as it grows.
func formatBytes(n int64) string {
	if n < 0 {
		return "?"
	}
	const unit = 1024
	if n < unit {
		return fmt.Sprintf("%d B", n)
	}
	div, exp := int64(unit), 0
	for v := n / unit; v >= unit && exp < 4; v /= unit {
		div *= unit
		exp++
	}
	value := float64(n) / float64(div)
	format := "%.1f %cB"
	if value >= 100 {
		format = "%.0f %cB"
	}
	return fmt.Sprintf(format, value, "KMGTP"[exp])
}

// formatSpeed renders bytes per second.
func formatSpeed(bps float64) string {
	if bps <= 0 {
		return "—"
	}
	return formatBytes(int64(bps)) + "/s"
}

// formatDuration renders a coarse, fixed-vocabulary duration. Sub-second
// precision would only add noise to a display that repaints ten times a
// second.
func formatDuration(d time.Duration) string {
	if d < 0 {
		return "—"
	}
	d = d.Round(time.Second)
	switch {
	case d < time.Minute:
		return fmt.Sprintf("%ds", int(d.Seconds()))
	case d < time.Hour:
		return fmt.Sprintf("%dm%02ds", int(d.Minutes()), int(d.Seconds())%60)
	default:
		return fmt.Sprintf("%dh%02dm", int(d.Hours()), int(d.Minutes())%60)
	}
}

// formatETA turns remaining bytes and a rate into a time, refusing to guess
// when either is unknown.
func formatETA(remaining int64, bps float64) string {
	if remaining <= 0 || bps <= 0 {
		return ""
	}
	return formatDuration(time.Duration(float64(remaining)/bps) * time.Second)
}

// barParts splits a progress bar into the filled run and the empty track, so
// each can be coloured separately — the contrast is what makes the bar
// readable, not the difference in stroke weight.
//
// A negative fraction means the total is unknown, which draws an empty track
// rather than a lie.
func barParts(fraction float64, width int) (fill, track string) {
	if width < 1 {
		return "", ""
	}
	if fraction < 0 {
		return "", strings.Repeat(trackRune, width)
	}
	if fraction > 1 {
		fraction = 1
	}
	filled := int(fraction*float64(width) + 0.5)
	return strings.Repeat(fillRune, filled), strings.Repeat(trackRune, width-filled)
}

// bar renders both parts uncoloured, which is what the tests measure.
func bar(fraction float64, width int) string {
	fill, track := barParts(fraction, width)
	return fill + track
}

// truncate shortens a string to a display width, marking the cut. Width is
// counted in runes, which is close enough for the names hosts hand out.
func truncate(s string, width int) string {
	if width <= 0 {
		return ""
	}
	runes := []rune(s)
	if len(runes) <= width {
		return s
	}
	if width == 1 {
		return "…"
	}
	return string(runes[:width-1]) + "…"
}

// pad right-pads to an exact rune width so columns line up.
func pad(s string, width int) string {
	if width <= 0 {
		return ""
	}
	s = truncate(s, width)
	if n := width - len([]rune(s)); n > 0 {
		return s + strings.Repeat(" ", n)
	}
	return s
}

// alignRight right-pads a field to an exact width, keeping columns in line
// down the screen. Overlong values are truncated rather than allowed to push
// the layout sideways.
func alignRight(s string, width int) string {
	if width <= 0 {
		return ""
	}
	s = truncate(s, width)
	if n := width - len([]rune(s)); n > 0 {
		return strings.Repeat(" ", n) + s
	}
	return s
}

File: frontend/src/format.ts
/** Human-readable byte count, e.g. "1.4 GB". */
export function formatBytes(bytes: number): string {
  if (!Number.isFinite(bytes) || bytes < 0) return '—';
  if (bytes < 1000) return `${bytes} B`;

  const units = ['kB', 'MB', 'GB', 'TB', 'PB'];
  let value = bytes / 1000;
  let unit = 0;
  while (value >= 1000 && unit < units.length - 1) {
    value /= 1000;
    unit += 1;
  }
  return `${value.toFixed(value < 10 ? 1 : 0)} ${units[unit]}`;
}

/** Transfer rate, e.g. "12.3 MB/s". */
export function formatSpeed(bytesPerSecond: number): string {
  if (!Number.isFinite(bytesPerSecond) || bytesPerSecond <= 0) return '—';
  return `${formatBytes(bytesPerSecond)}/s`;
}

/** Remaining time from the current rate, e.g. "2m 30s". */
export function formatEta(remaining: number, bytesPerSecond: number): string {
  if (remaining <= 0 || bytesPerSecond <= 0) return '—';
  return formatDuration(remaining / bytesPerSecond);
}

/** Duration in seconds as a compact string. */
export function formatDuration(seconds: number): string {
  if (!Number.isFinite(seconds) || seconds < 0) return '—';
  if (seconds < 60) return `${Math.round(seconds)}s`;

  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m ${Math.round(seconds % 60)}s`;

  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ${minutes % 60}m`;
  return `${Math.floor(hours / 24)}d ${hours % 24}h`;
}

/**
 * Completion as a 0-100 percentage. Returns null when the total is unknown,
 * so callers can show an indeterminate bar rather than a misleading zero.
 */
export function percentOf(done: number, total: number): number | null {
  if (total <= 0) return null;
  return Math.min(100, Math.max(0, (done / total) * 100));
}

/** Short label for a host, e.g. "gofile". */
export function hostLabel(host: string): string {
  return host || 'direct';
}
